internal/client/block_rules.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2025 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package client
+
+import "github.com/ChainSafe/gossamer/internal/primitives/runtime"
+
+type BadBlockData[H runtime.Hash, N runtime.Number] struct {
+	number N
+	hash   H
+}
+
+type BlockRules[H runtime.Hash, N runtime.Number] struct {
+	bad   map[H]struct{} // Set with bad blocks
+	forks map[N]H
+}
+
+// Chain specification rules lookup result.
+type LookupResult interface {
+	isLookupResult()
+}
+
+type (
+	// Specification rules do not contain any special rules about this block
+	LookupResultNotSpecial struct{}
+	// The block is known to be bad and should not be imported
+	LookupResultKnownBad struct{}
+	// There is a specified canonical block hash for the given height
+	LookupResultExpected[H runtime.Hash] struct {
+		hash H
+	}
+)
+
+func (LookupResultNotSpecial) isLookupResult()  {}
+func (LookupResultKnownBad) isLookupResult()    {}
+func (LookupResultExpected[H]) isLookupResult() {}
```

Comment:
nit: I think moving the type closer to the construct/receiver methods is better.
```suggestion
// Chain specification rules lookup result.
type LookupResult interface {
	isLookupResult()
}

type (
	// Specification rules do not contain any special rules about this block
	LookupResultNotSpecial struct{}
	// The block is known to be bad and should not be imported
	LookupResultKnownBad struct{}
	// There is a specified canonical block hash for the given height
	LookupResultExpected[H runtime.Hash] struct {
		hash H
	}
)

func (LookupResultNotSpecial) isLookupResult()  {}
func (LookupResultKnownBad) isLookupResult()    {}
func (LookupResultExpected[H]) isLookupResult() {}

type BlockRules[H runtime.Hash, N runtime.Number] struct {
	bad   map[H]struct{} // Set with bad blocks
	forks map[N]H
}
```